crates/crypto/src/key_manager/local_key_manager.rs

```diff
@@ -0,0 +1,127 @@
+use crate::key::{KeyType, PrivateKey, PublicKey};
+use crate::key_manager::key_store::{InMemoryKeyStore, KeyStore};
+use crate::key_manager::{KeyManager, KeyManagerError};
+use ssi_jwk::JWK;
+use std::sync::Arc;
+
+/// Implementation of the [`KeyManager`] trait with key generation local to the device/platform it
+/// is being run. Key storage is provided by a [`KeyStore`] trait implementation, allowing the keys
+/// to be stored wherever is most appropriate for the application.
+pub struct LocalKeyManager {
+    key_store: Arc<dyn KeyStore>,
```

Comment:
Arc is a shared data reference that uses the concept of reference counting. When you create an Arc via `Arc::new()` around some piece of data, that data is placed on the heap along with two counters for "strong" and "weak" references to it. After `Arc::new()`, it looks something like this:

![Untitled scene](https://github.com/TBD54566975/web5-rs/assets/88001738/31fbf4cb-204a-400f-8712-15714ca1b2b6)

The `Arc` you get back is actually a pointer to some heap allocated memory (called `ArcInner`), which contains the counts, as well as the actual shared data itself. If the `Arc` pointer you get back goes out of scope and gets deallocated, the strong reference counter that the `ArcInner` was pointing to is atomically decremented. When the strong count falls to zero, the heap allocated memory that contains the actual data is then cleaned up.

Now looking at these particular lines:

```
pub struct LocalKeyManager {
    key_store: Arc<dyn KeyStore>,
}
```

What this is saying is that `LocalKeyManager` owns an `Arc`, which points to a KeyStore. The lifetime of the Arc will only end when the `LocalKeyManager`'s lifetime ends. This means that the strong reference count will remain non-zero for the duration of `LocalKeyManager`'s lifetime, guaranteeing that it's always available to use.

In order to construct a `LocalKeyManager`, the caller must to provide an `Arc<dyn Keystore>` value in the constructor. This can be done either by creating a new Arc like I mentioned earlier with `Arc::new()`, or the caller can clone an existing Arc.

Cloning an Arc isn't actually cloning the heap allocated data. Instead, it's cloning the pointer to that heap allocated data. As a part of the clone operation, the strong reference count is atomically incremented. So, even if the originally cloned Arc goes out of scope and gets deallocated, decrementing the reference count, the reference count is still non-zero.

---

I hope that made some sort of sense 😅 . Happy to chat through this more if you'd like more clarity, or had additional questions! Arc is definitely one of the more interesting data types in Rust, and it's heavily used throughout the bindings generation process which will come in future PRs.

Reply:
Thanks! Asked more in the line below.